src/derivative.jl

```diff
@@ -0,0 +1,38 @@
+default_derivative_algorithm() = "exact"
+
+function derivative(
+  ψ::AbstractITensorNetwork, vertices::Vector; alg=default_derivative_algorithm(), kwargs...
+)
+  return derivative(Algorithm(alg), ψ, vertices; kwargs...)
+end
+
+function derivative(
+  ::Algorithm"exact",
+  ψ::AbstractITensorNetwork,
+  vertices::Vector;
+  contraction_sequence_alg="optimal",
+  kwargs...,
+)
+  ψ_reduced = Vector{ITensor}(subgraph(ψ, vertices))
+  sequence = contraction_sequence(ψ_reduced; alg=contraction_sequence_alg)
+  return ITensor[contract(ψ_reduced; sequence, kwargs...)]
+end
+
+function derivative(
+  ::Algorithm"bp",
+  ψ::AbstractITensorNetwork,
+  verts::Vector;
+  (bp_cache!)=nothing,
+  update_bp_cache=isnothing(bp_cache!),
+  bp_cache_update_kwargs=default_cache_update_kwargs(bp_cache!),
+)
+  if isnothing(bp_cache!)
+    bp_cache! = Ref(BeliefPropagationCache(ψ))
+  end
+
+  if update_bp_cache
+    bp_cache![] = update(bp_cache![]; bp_cache_update_kwargs...)
+  end
+
+  return incoming_messages(bp_cache![], setdiff(vertices(ψ), verts))
```

Comment:
In general, shouldn't the derivative include tensors both from the factor and from the messages? I.e. to build a projected MPO, we need both the environment tensors and the MPO tensors in the region we are updating.

Reply:
Yes, `incoming_messages` automatically takes care of this in the BP code and will include the tensors on the vertices from the region which are not specified in `verts`. Obviously that maybe suggests `incoming_messages` is a misnomer. So if you think there might be a more appropriate name like `environment_tensors` I am happy to revert to that

Reply:
Yeah, we should think of a better name for that functionality, `incoming_messages` is misleading. `environment_tensors` or just `environment` seems reasonable. I think I prefer `environment`, we can see if that becomes confusing or unclear.

Reply:
Okay let me go with that!